src/App.tsx

```diff
@@ -1,26 +1,259 @@
+        }),
+      ),
+    )
+      .then(() => {
+        setTodos(updatedTodos);
+      })
+      .finally(() => setIsLoading(false));
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    const deletedTodo = todos.find(todo => todo.id === todoId);
+
+    if (!deletedTodo) {
+      return;
+    }
+
+    setIsLoading(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+      })
+      .catch(() => {
+        setError('Unable to delete a todo');
+      })
+      .finally(() => setIsLoading(false));
+  };
+
+  const handleEditTodo = (todo: Todo) => {
+    setEditingTodoId(todo.id);
+    setEditingTodoTitle(todo.title);
+  };
+
+  const handleUpdateTodo = (event: React.FormEvent, todoId: number) => {
+    event.preventDefault();
+    if (editingTodoTitle.trim() === '') {
+      handleDeleteTodo(todoId);
+
+      return;
+    }
+
+    const updatedTodo = todos.find(todo => todo.id === editingTodoId);
+
+    if (!updatedTodo) {
+      return;
+    }
+
+    if (updatedTodo.title === editingTodoTitle) {
+      setEditingTodoId(null);
+
+      return;
+    }
+
+    setIsLoading(true);
+
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    const newTodo = { ...updatedTodo, title: editingTodoTitle };
+
+    updateTodo(todoId, newTodo)
+      .then(() => {
+        setTodos(prevTodos =>
+          prevTodos.map(todo => (todo.id === todoId ? newTodo : todo)),
+        );
+        setEditingTodoId(null);
+      })
+      .catch(() => {
+        setError('Unable to update a todo');
+      })
+      .finally(() => setIsLoading(false));
+  };
+
+  const handleCancelEdit = () => {
+    setEditingTodoId(null);
+  };
+
+  const handleClearCompleted = () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    if (completedTodos.length === 0) {
+      return;
+    }
+
+    setIsLoading(true);
+
+    Promise.all(
+      completedTodos.map(todo =>
+        deleteTodo(todo.id).catch(() => {
+          setError('Unable to delete a todo');
+        }),
+      ),
+    )
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => !todo.completed));
+      })
+      .finally(() => setIsLoading(false));
+  };
+
+  const allCompleted = todos.length > 0 && todos.every(todo => todo.completed);
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">
-          React Todo App - Load Todos
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      {!USER_ID ? (
+        <UserWarning />
+      ) : (
+        <>
+          <h1 className="todoapp__title">todos</h1>
+
+          <div className="todoapp__content">
+            {todos.length === 0 ? (
```

Comment:
```suggestion
            {!todos.length ? (
```

Reply:
Not fixed:
![image](https://github.com/user-attachments/assets/9e9f8a8e-c6dc-4569-b82b-74ec86489842)